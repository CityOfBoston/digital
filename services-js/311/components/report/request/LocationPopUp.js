// @flow

import React from 'react';
import { css } from 'glamor';
import { observable, action, runInAction } from 'mobx';
import { observer } from 'mobx-react';

import type { AppStore } from '../../../data/store';

import { SMALL_SCREEN } from '../../style-constants';

const CONTENT_STYLE = css({
  display: 'flex',
  flexDirection: 'column',
  alignItems: 'stretch',
  [SMALL_SCREEN]: {
    marginTop: 0,
  },
});

const BUTTON_ROW_STYLE = css({
  alignItems: 'center',
});

export type Props = {
  store: AppStore,
  addressSearch: ?((string) => Promise<boolean>),
  nextFunc: () => void,
}

@observer
export default class LocationPopUp extends React.Component {
  props: Props;
  @observable addressQuery: string = '';

  @action.bound
  whenSearchInput(ev: SyntheticInputEvent) {
    this.addressQuery = ev.target.value;
  }

  whenSearchSubmit = async (ev: SyntheticInputEvent) => {
    const { addressSearch } = this.props;

    ev.preventDefault();

    if (!addressSearch) {
      return;
    }

    const found = await addressSearch(this.addressQuery);
    if (found) {
      runInAction('whenSearchSubmit success', () => {
        this.addressQuery = '';
      });
    }
  }

  render() {
    const { store, nextFunc, addressSearch } = this.props;
    const { address, requirementsMet, required } = store.locationInfo;

    return (
      <div className={CONTENT_STYLE}>
        <h3 className="t--info">Where is the problem happening?</h3>

        <hr className="hr hr--dash" />

        { address && <div className="addr addr--s" style={{ whiteSpace: 'pre-line' }}>{ address }</div> }

        <form className="sf sf--sm sf--y m-v300" onSubmit={this.whenSearchSubmit}>
          <div className="sf-i">
            <input
              className="sf-i-f"
              onInput={this.whenSearchInput}
              value={this.addressQuery}
              placeholder={address ? 'Search for another address…' : 'Search for an address…'}
              type="text"
            />

            <button className="sf-i-b" type="submit" disabled={this.addressQuery.length === 0 || !addressSearch}>Search</button>
          </div>
        </form>

        <div className={`g ${BUTTON_ROW_STYLE.toString()}`}>
          <div className="g--6">
            { !required && <a href="javascript:void(0)" onClick={nextFunc}>Continue without location</a> }
          </div>
          <button className="btn g--6" disabled={!requirementsMet} onClick={nextFunc}>Next</button>
        </div>
      </div>
    );
  }
}
