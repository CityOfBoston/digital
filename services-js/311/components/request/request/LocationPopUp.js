// @flow

import React from 'react';
import { css } from 'emotion';
import { action, reaction } from 'mobx';
import { observer } from 'mobx-react';

import type { AppStore } from '../../../data/store';
import type AddressSearch from '../../../data/store/AddressSearch';
import type { SearchAddressPlace, AddressUnit } from '../../../data/types';
import type RequestForm from '../../../data/store/RequestForm';

import LoadingIcons from '../../common/LoadingIcons';
import { LocationMapWithLibrary } from '../../map/LocationMap';
import {
  MEDIA_LARGE,
  HEADER_HEIGHT,
  GRAY_000,
  GRAY_200,
  CHARLES_BLUE,
} from '../../style-constants';

import waypointMarkers, {
  WAYPOINT_NUMBER_STYLE,
} from '../../map/WaypointMarkers';

// On large screen sizes below, the location picker is in a little
// dialog that floats over the map. We let it size itself automatically
// from its content. On mobile devices, the picker has a fixed height
// and we use flex: 1 to fill that height, so that the button at the
// bottom has a consistent position.
//
// We have to disable the flex for desktop because of IE, which renders
// the flexed children with 0 height, rather than letting their
// intrinsic heights grow the flex container.

const CONTENT_STYLE = css({
  display: 'flex',
  flexDirection: 'column',
  alignItems: 'stretch',
  // We can't flex: 1 to the min-height of the mobile FormDialog in IE
  // 0.44rem is the border top of the dialog
  minHeight: `calc(100vh - ${HEADER_HEIGHT}px - 0.44rem)`,
  [MEDIA_LARGE]: {
    minHeight: 'auto',
    flex: 'none',
  },
});

const FORM_WRAPPER_STYLE = css({
  display: 'flex',
  flexDirection: 'column',
  flexGrow: 1,
  flexShrink: 0,
  overflowY: 'auto',

  [MEDIA_LARGE]: {
    flex: 'none',
    marginTop: '0.25em',
    overflowY: 'visible',
  },
});

const ADDRESS_WRAPPER_STYLE = css({
  flex: 1,
  [MEDIA_LARGE]: {
    flex: 'none',
  },
});

const BUTTON_ROW_STYLE = css({
  alignItems: 'center',
});

const MAP_CONTAINER_STYLE = css({
  height: '38vh',
  transition: 'height 200ms',
});

const SMALL_MAP_CONTAINER_STYLE = css({
  height: '15vh',
  minHeight: 130,
  transition: 'height 200ms',
});

const ADDRESS_LIST_STYLE = css({
  [MEDIA_LARGE]: {
    maxHeight: 220,
    overflowY: 'auto',
  },
});

const ADDRESS_ROW_STYLE = css({
  textAlign: 'left',
  fontSize: 'inherit',
  width: '100%',
  borderLeft: 'none',
  borderRight: 'none',
  display: 'block',
  position: 'relative',
  outlineOffset: -3,
  outlineWidth: 3,
  color: CHARLES_BLUE,
  ':hover': {
    color: CHARLES_BLUE,
  },
  paddingRight: 94,
  whiteSpace: 'nowrap',
});

const UNIT_PICKER_STYLE = css({
  position: 'absolute',
  zIndex: 2,
  right: 10,
  top: 10,
  '& .sel-c:after': {
    content: '"Units"',
  },
});

const BORDER_TOP_ADDRESS_STYLE = css({
  borderTop: `1.5px dashed ${GRAY_200}`,
});

const NO_BORDER_TOP_ADDRESS_STYLE = css({
  borderTop: `1.5px dashed transparent`,
});

const BORDER_BOTTOM_ADDRESS_STYLE = css({
  borderBottom: `1.5px dashed ${GRAY_200}`,
});

const NO_BORDER_BOTTOM_ADDRESS_STYLE = css({
  borderBottom: `1.5px dashed transparent`,
});

const SELECTED_ADDRESS_STYLE = css(NO_BORDER_TOP_ADDRESS_STYLE, {
  borderColor: 'transparent',
  backgroundColor: GRAY_000,
});

// Needed to reset <button> default styles
const NOT_SELECTED_ADDRESS_STYLE = css({
  background: 'transparent',
});

const WAYPOINT_WRAPPER_STYLE = css({
  verticalAlign: 'middle',
  display: 'inline-block',
  position: 'relative',
  marginRight: '.825em',
});

const WAYPOINT_STYLE = css({
  width: 27,
  height: 50,
});

const SEARCH_RESULTS_HEADER_STYLE = css({
  fontSize: 14,
  color: GRAY_200,
  letterSpacing: '1px',
  paddingBottom: 0,
  paddingTop: 0,
  marginBottom: '.475em',
});

const LOADING_CONTAINER_STYLE = css({ height: 85, display: 'flex' });

export type Props = {|
  store: AppStore,
  requestForm: RequestForm,
  nextFunc: () => mixed,
  nextIsSubmit: boolean,
|};

@observer
export default class LocationPopUp extends React.Component<Props> {
  containerEl: ?HTMLElement;
  searchField: ?HTMLInputElement;

  updateFormFromMapDisposer: Function;

  constructor(props: Props) {
    super(props);
  }

  @action
  componentWillMount() {
    const { store: { addressSearch, accessibility }, requestForm } = this.props;

    addressSearch.query = '';

    this.updateFormFromMapDisposer = reaction(
      () => ({
        location: addressSearch.location,
        address: addressSearch.address,
        addressId: addressSearch.addressId,
        intent: addressSearch.intent,
      }),
      ({ location, address, addressId, intent }) => {
        requestForm.location = location;
        requestForm.address = address;
        requestForm.addressId = addressId;
        requestForm.addressIntent = intent;

        if (requestForm.address) {
          accessibility.message = `Selected address: ${requestForm.address}`;
          accessibility.interrupt = true;
        }
      },
      {
        fireImmediately: true,
        name: 'update form from map',
      }
    );
  }

  componentWillUnmount() {
    this.updateFormFromMapDisposer();
  }

  setContainer = (containerEl: ?HTMLElement) => {
    this.containerEl = containerEl;
  };

  setSearchField = (searchField: ?HTMLInputElement) => {
    this.searchField = searchField;
  };

  @action.bound
  whenSearchInput(ev: SyntheticInputEvent<>) {
    const { store: { addressSearch } } = this.props;
    const query = ev.target.value;

    addressSearch.query = query;

    if (query === '') {
      addressSearch.clearPlaces();
    }
  }

  @action.bound
  whenSearchSubmit(ev: SyntheticInputEvent<>) {
    ev.preventDefault();

    const { store: { ui: { belowMediaLarge }, addressSearch } } = this.props;

    if (this.containerEl) {
      // we update this on search because when the component is mounted it's
      // still being animated down to its small width.
      addressSearch.searchPopupWidth = this.containerEl.getBoundingClientRect().right;
    }

    // on mobile, select the first result by default
    addressSearch.search(belowMediaLarge);
  }

  @action.bound
  continueWithLocation() {
    const { nextFunc, requestForm } = this.props;
    requestForm.sendLocation = true;
    nextFunc();
  }

  @action.bound
  continueWithoutLocation() {
    const { nextFunc, requestForm } = this.props;
    requestForm.sendLocation = false;
    nextFunc();
  }

  render() {
    const {
      requestForm,
      nextIsSubmit,
      store: { addressSearch, ui },
    } = this.props;
    const { belowMediaLarge } = ui;
    const { locationRequirementsMet } = requestForm;
    const {
      places,
      lastSearchError,
      searching,
      query,
      location,
    } = addressSearch;

    return (
      <div className={CONTENT_STYLE}>
        {this.maybeRenderMap()}

        <div
          className={`${FORM_WRAPPER_STYLE.toString()}`}
          ref={this.setContainer}
        >
          <div className="p-a300">
            <form
              className="sf sf--md sf--y"
              onSubmit={this.whenSearchSubmit}
              action="#"
            >
              <div className="sf-i">
                <input
                  className="sf-i-f"
                  onChange={this.whenSearchInput}
                  value={query}
                  ref={this.setSearchField}
                  aria-label="Address search field"
                  placeholder={'Search address…'}
                  type="text"
                />

                <button
                  className="sf-i-b"
                  type="submit"
                  disabled={query.length === 0}
                >
                  Search
                </button>
              </div>
            </form>
          </div>

          <div className={ADDRESS_WRAPPER_STYLE}>
            {searching && this.renderLoading()}
            {lastSearchError && !location && this.renderSearchError()}
            {lastSearchError && location && this.renderGeocodeError()}

            {places &&
              (places.length > 0
                ? this.renderAddressList(places)
                : this.renderNotFound())}
          </div>

          <div className={`g p-a300 ${BUTTON_ROW_STYLE.toString()}`}>
            {!belowMediaLarge && this.renderContinueWithoutLocation()}
            <button
              type="button"
              className="btn g--5"
              disabled={!locationRequirementsMet}
              onClick={this.continueWithLocation}
            >
              {nextIsSubmit ? 'Submit' : 'Next'}
            </button>
            {belowMediaLarge && this.renderContinueWithoutLocation()}
          </div>
        </div>
      </div>
    );
  }

  renderContinueWithoutLocation() {
    const { requestForm: { locationRequired } } = this.props;

    return (
      <div className="g--7 t--subinfo m-v200">
        {!locationRequired &&
          <a href="javascript:void(0)" onClick={this.continueWithoutLocation}>
            Continue without location
          </a>}
      </div>
    );
  }

  maybeRenderMap() {
    const { store } = this.props;
    const { belowMediaLarge } = store.ui;
    const { places } = store.addressSearch;

    if (!belowMediaLarge) {
      return null;
    }

    const smallPicker = places && places.length > 1;

    return (
      <div
        className={`${smallPicker
          ? SMALL_MAP_CONTAINER_STYLE.toString()
          : MAP_CONTAINER_STYLE.toString()}`}
      >
        <LocationMapWithLibrary
          store={store}
          mode="picker"
          mobile
          smallPicker={!!smallPicker}
        />
      </div>
    );
  }

  renderLoading() {
    return (
      <div className={`${LOADING_CONTAINER_STYLE.toString()}`} key="loading">
        <LoadingIcons />
      </div>
    );
  }

  renderSearchError() {
    return (
      <div className="p-a300" key="error" style={{ paddingTop: 0 }}>
        <p className="t--err t--info" style={{ marginTop: 0 }}>
          We couldn’t find that address because of a server problem.
        </p>
        <p className="t--subinfo">
          Please try again later, or choose a location using the map.
        </p>
      </div>
    );
  }

  renderGeocodeError() {
    return (
      <div className="p-a300" key="error" style={{ paddingTop: 0 }}>
        <p className="t--err t--info" style={{ marginTop: 0 }}>
          We couldn’t find an address for the map pin because of a server
          problem.
        </p>
        <p className="t--subinfo">
          If your case is for something at a specific street address, please try
          again later.
        </p>
      </div>
    );
  }

  renderNotFound() {
    const { location, query } = this.props.store.addressSearch;

    return (
      <div className="p-a300 t--info" key="not-found">
        <span className="t--err">
          {location
            ? 'Please pick a location within Boston'
            : `We could not find an address or intersection in Boston for “${query}”`}
        </span>
      </div>
    );
  }

  renderAddressList(places: Array<SearchAddressPlace>) {
    const { addressSearch } = this.props.store;
    const { lastQuery } = addressSearch;

    return (
      <div className="" key="places">
        {!!lastQuery &&
          places.length !== 1 &&
          <div
            className={`t--sans t--upper t--g300 p-a300 t--ellipsis ${SEARCH_RESULTS_HEADER_STYLE.toString()}`}
          >
            {places.length} results for {lastQuery}
          </div>}

        <div className={`${ADDRESS_LIST_STYLE.toString()}`}>
          {places.map((place, idx) =>
            <AddressRow
              place={place}
              index={idx}
              key={place.addressId}
              placeCount={places.length}
              addressSearch={addressSearch}
            />
          )}
        </div>
      </div>
    );
  }
}

type AddressRowProps = {
  place: SearchAddressPlace,
  index: number,
  placeCount: number,
  addressSearch: AddressSearch,
};

@observer
class AddressRow extends React.Component<AddressRowProps> {
  @action.bound
  selectAddress() {
    const { index, addressSearch } = this.props;

    addressSearch.highlightedPlaceIndex = index;

    if (index !== addressSearch.currentPlaceIndex) {
      addressSearch.currentPlaceIndex = index;
      addressSearch.currentUnitIndex = 0;
    }
  }

  @action.bound
  highlightAddress() {
    const { index, addressSearch } = this.props;

    addressSearch.highlightedPlaceIndex = index;
  }

  @action.bound
  whenUnitChange(ev: SyntheticInputEvent<>) {
    const { addressSearch } = this.props;
    addressSearch.currentUnitIndex = parseInt(ev.target.value, 10);
  }

  render() {
    const {
      index,
      place,
      placeCount,
      addressSearch: { address, highlightedPlaceIndex, currentPlaceIndex },
    } = this.props;

    const highlighted = index === highlightedPlaceIndex;
    const current = index === currentPlaceIndex;

    const classes = [
      ADDRESS_ROW_STYLE,
      'p-a300',
      current ? SELECTED_ADDRESS_STYLE : NOT_SELECTED_ADDRESS_STYLE,
      current || index === currentPlaceIndex + 1
        ? NO_BORDER_TOP_ADDRESS_STYLE
        : BORDER_TOP_ADDRESS_STYLE,
      index === placeCount - 1 && index !== currentPlaceIndex
        ? BORDER_BOTTOM_ADDRESS_STYLE.toString()
        : NO_BORDER_BOTTOM_ADDRESS_STYLE.toString(),
    ];

    return (
      <div style={{ position: 'relative' }} key={place.addressId}>
        <button
          type="button"
          className={classes.join(' ')}
          onClick={this.selectAddress}
          onMouseEnter={this.highlightAddress}
          onMouseLeave={this.highlightAddress}
        >
          <div className={`${WAYPOINT_WRAPPER_STYLE.toString()}`}>
            <div
              className={`${WAYPOINT_STYLE.toString()}`}
              dangerouslySetInnerHTML={{
                __html:
                  waypointMarkers[highlighted ? 'greenFilled' : 'orangeFilled']
                    .html,
              }}
            />
            {placeCount > 1 &&
              <div className={`t--sans ${WAYPOINT_NUMBER_STYLE.toString()}`}>
                {index + 1}
              </div>}
          </div>
          <div
            className="addr addr--s"
            style={{
              whiteSpace: 'pre-line',
              display: 'inline-block',
              verticalAlign: 'middle',
              lineHeight: 1.4,
            }}
          >
            {!place.exact &&
              <div className="t--subinfo tt-n">Approximate location</div>}
            {(current
              ? address
              : place.units && place.units.length
                ? place.units[0].address
                : place.address).trim()}
          </div>

          <div style={{ clear: 'both' }} />
        </button>

        {current &&
          place.units.length > 1 &&
          this.renderUnitSelector(place.units)}
      </div>
    );
  }

  renderUnitSelector(units: Array<AddressUnit>) {
    const { currentUnitIndex } = this.props.addressSearch;

    return (
      <div className={`${UNIT_PICKER_STYLE.toString()} sel`}>
        <label className="sel-l sel-l--sq" htmlFor="unit-menu">
          Apartment or Unit
        </label>
        <div className="sel-c sel-c--sq">
          <select
            id="unit-menu"
            className="sel-f sel-f--sq"
            value={currentUnitIndex}
            onChange={this.whenUnitChange}
          >
            {units.map(({ addressId, streetAddress }, idx) =>
              <option value={idx} key={addressId}>
                {streetAddress}
              </option>
            )}
          </select>
        </div>
      </div>
    );
  }
}
