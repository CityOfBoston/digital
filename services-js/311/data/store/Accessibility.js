// @flow

import { observable, autorun } from 'mobx';

export default class Accessibility {
  el: ?HTMLElement = null;

  @observable message: string = '';
  @observable interrupt: boolean = false;

  messageListenerDisposer: ?Function = null;

  attach() {
    this.el = document.getElementById('ariaLive');

    this.messageListenerDisposer = autorun(
      () => {
        if (this.el) {
          this.el.innerText = this.message;
          this.el.setAttribute(
            'aria-live',
            this.interrupt ? 'assertive' : 'polite'
          );
        }
      },
      {
        name: 'a11y message listener',
      }
    );
  }

  detatch() {
    this.el = null;

    if (this.messageListenerDisposer) {
      this.messageListenerDisposer();
    }
  }
}
