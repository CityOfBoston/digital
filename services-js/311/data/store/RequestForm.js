// @flow

import { observable, computed, action } from 'mobx';

import type { Service } from '../types';
import Question from './Question';

export class ContactInfo {
  @observable required: boolean = false;
  @observable firstName: string = '';
  @observable lastName: string = '';
  @observable email: string = '';
  @observable phone: string = '';

  // can be false even when required is false to differentiate between submitting
  // and skipping
  @computed
  get requirementsMet(): boolean {
    return !!(this.firstName && this.lastName && this.email);
  }
}

export class LocationInfo {
  @observable required: boolean = false;
  @observable address: string = '';
  @observable.shallow location: ?{ lat: number, lng: number } = null;

  // can be false even when required is false to differentiate between submitting
  // and skipping
  @computed
  get requirementsMet(): boolean {
    return this.address.length > 0;
  }
}

// Class to encapsulate the request that's being built up over the course of
// the web flow.
export default class RequestForm {
  @observable description: string = '';
  @observable mediaUrl: string = ''

  @observable contactInfo: ContactInfo = new ContactInfo();
  @observable locationInfo: LocationInfo = new LocationInfo();
  @observable questions: Question[] = [];

  @action
  updateForService(service: ?Service) {
    this.questions = [];

    if (service) {
      this.contactInfo.required = service.contactRequired;
      this.locationInfo.required = service.locationRequired;
      this.questions = Question.buildQuestions(service.attributes);
    }
  }

  @computed get questionRequirementsMet(): boolean {
    return this.questions.filter((q) => !q.required || !q.visible || q.requirementsMet).length === this.questions.length;
  }
}
