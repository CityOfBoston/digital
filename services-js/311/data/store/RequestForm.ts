import { observable, computed } from 'mobx';

import { Service } from '../types';
import Question from './Question';

// Class to encapsulate the request that's being built up over the course of
// the web flow.
export default class RequestForm {
  private readonly service: Service | null;

  @computed
  get code(): string {
    return this.service ? this.service.code : '';
  }

  // This is the description originally typed into the home page. We keep track
  // of it to help train the classifier.
  descriptionForClassifier: string = '';
  @observable description: string = '';
  @observable mediaUrl: string = '';

  @observable address: string = '';
  @observable.shallow location: { lat: number; lng: number } | null = null;
  @observable addressId: string | null = null;
  @observable addressIntent: 'ADDRESS' | 'LATLNG' = 'ADDRESS';
  @observable sendLocation: boolean = false;
  @observable locationRequirementsMet: boolean = false;

  @observable firstName: string = '';
  @observable lastName: string = '';
  @observable email: string = '';
  @observable phone: string = '';
  @observable sendContactInfo: boolean = false;

  @observable questions: Question[] = [];

  @computed
  get showContactInfoForm(): boolean {
    return !!this.service && this.service.contactRequirement !== 'HIDDEN';
  }

  @computed
  get contactInfoRequired(): boolean {
    return !!this.service && this.service.contactRequirement === 'REQUIRED';
  }

  // can be false even when required is false to differentiate between submitting
  // and skipping
  @computed
  get contactInfoRequirementsMet(): boolean {
    return !!(this.firstName && this.lastName && this.email);
  }

  @computed
  get showLocationPicker(): boolean {
    return !!this.service && this.service.locationRequirement !== 'HIDDEN';
  }

  @computed
  get locationRequired(): boolean {
    return !!this.service && this.service.locationRequirement === 'REQUIRED';
  }

  @computed
  get questionRequirementsMet(): boolean {
    return (
      this.questions.filter(q => !q.required || !q.visible || q.requirementsMet)
        .length === this.questions.length
    );
  }

  constructor(service: Service | null = null) {
    this.service = service;
    if (service) {
      this.questions = Question.buildQuestions(service.attributes);
    } else {
      this.questions = [];
    }
  }
}
