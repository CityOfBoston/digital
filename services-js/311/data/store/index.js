// @flow

import { observable, computed, useStrict } from 'mobx';
import type { IPromiseBasedObservable } from 'mobx-utils';

import type { Service, ServiceSummary, SubmittedRequest } from '../types';
import Question from './Question';

// MobX will enforce that state changes only happen in action blocks.
useStrict(true);

export class ContactInfo {
  @observable required: boolean = false;
  @observable firstName: string = '';
  @observable lastName: string = '';
  @observable email: string = '';
  @observable phone: string = '';

  // can be false even when required is false to differentiate between submitting
  // and skipping
  @computed
  get requirementsMet(): boolean {
    return !!(this.firstName && this.lastName);
  }
}

export class LocationInfo {
  @observable required: boolean = false;
  @observable address: string = '';
  @observable.shallow location: ?{| lat: number, lng: number |} = null;

  // can be false even when required is false to differentiate between submitting
  // and skipping
  @computed
  get requirementsMet(): boolean {
    return this.address.length > 0;
  }
}

export class AppStore {
  @observable description: string = '';

  @observable contactInfo: ContactInfo = new ContactInfo();
  @observable locationInfo: LocationInfo = new LocationInfo();
  @observable questions: Question[] = [];

  @observable.shallow serviceSummaries: ServiceSummary[];
  serviceCache: Map<string, Service> = observable.shallowMap({});

  apiKeys: {[service: string]: string} = {};
  isPhone: boolean = false;

  @observable requestSubmission: ?IPromiseBasedObservable<SubmittedRequest> = null;

  @observable.ref _currentService: ?Service = null;
  @observable.ref currentServiceError: ?Object = null;

  @computed get currentService(): ?Service {
    return this._currentService;
  }

  set currentService(service: ?Service) {
    if (service === this._currentService) {
      return;
    }

    this._currentService = service;
    this.currentServiceError = null;
    this.questions = [];

    if (service) {
      try {
        this.contactInfo.required = service.contactRequired;
        this.locationInfo.required = service.locationRequired;
        this.questions = Question.buildQuestions(service.attributes);
      } catch (e) {
        this.currentServiceError = e;
      }
    }
  }

  @computed
  get questionRequirementsMet(): boolean {
    return this.questions.filter((q) => !q.required || !q.visible || q.requirementsMet).length === this.questions.length;
  }
}

let browserStore: AppStore;

/**
 * Returns a Redux store for this application. For the server, always returns a
 * fresh store with the given initialState. On the browser, will always return
 * the same store and initialState is ignored after the first call.
 */
export default function getStore(): AppStore {
  if (process.browser && browserStore) {
    return browserStore;
  }

  const store = new AppStore();

  if (process.browser) {
    browserStore = store;
  }

  return store;
}
