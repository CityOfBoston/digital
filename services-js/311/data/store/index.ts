/* global window */

import { observable, configure as mobxConfigure, action } from 'mobx';
import Router from 'next/router';

import RequestSearch from './RequestSearch';
import AddressSearch from './AddressSearch';
import Ui from './Ui';
import BrowserLocation from './BrowserLocation';
import AllServices from './AllServices';
import {
  FetchGraphql,
  ScreenReaderSupport,
  GaSiteAnalytics,
  RouterListener,
} from '@cityofboston/next-client-common';
import { SiteAnalytics } from '@cityofboston/next-client-common/build/SiteAnalytics';

// MobX will enforce that state changes only happen in action blocks.
mobxConfigure({ enforceActions: true });

export type LanguagePreference = {
  code: string;
  region: string | null;
  quality: number;
};

export class AppStore {
  screenReaderSupport: ScreenReaderSupport = new ScreenReaderSupport();
  requestSearch: RequestSearch = new RequestSearch();
  ui: Ui = new Ui();
  browserLocation: BrowserLocation = new BrowserLocation();
  addressSearch: AddressSearch = new AddressSearch();
  allServices: AllServices = new AllServices();
  siteAnalytics: SiteAnalytics = new GaSiteAnalytics();

  languages: LanguagePreference[] = [];

  @observable
  liveAgentAvailable: boolean = (typeof window !== 'undefined' &&
    window.LIVE_AGENT_AVAILABLE) ||
  false;

  _liveAgentButtonId: string = '';

  get liveAgentButtonId(): string {
    return this._liveAgentButtonId;
  }

  set liveAgentButtonId(liveAgentButtonId: string) {
    this._liveAgentButtonId = liveAgentButtonId;

    if (typeof window !== 'undefined') {
      if (window.liveagent) {
        this.listenForLiveAgentEvents(liveAgentButtonId);
      } else {
        if (!window._laq) {
          window._laq = [];
        }

        window._laq.push(() =>
          this.listenForLiveAgentEvents(liveAgentButtonId)
        );
      }
    }
  }

  listenForLiveAgentEvents(liveAgentButtonId: string) {
    window.liveagent!.addButtonEventHandler(
      liveAgentButtonId,
      this.liveAgentEventHandler
    );
  }

  @action.bound
  liveAgentEventHandler(event: string) {
    switch (event) {
      case 'BUTTON_AVAILABLE':
        this.liveAgentAvailable = true;
        break;
      case 'BUTTON_UNAVAILABLE':
        this.liveAgentAvailable = false;
        break;
      default:
        break;
    }
  }
}

let browserStore: AppStore;

/**
 * Returns a Redux store for this application. For the server, always returns a
 * fresh store with the given initialState. On the browser, will always return
 * the same store and initialState is ignored after the first call.
 */
export default function getStore(
  fetchGraphql: FetchGraphql | null = null
): AppStore {
  if (process.browser && browserStore) {
    return browserStore;
  }

  const store = new AppStore();

  if (process.browser) {
    browserStore = store;
    store.screenReaderSupport.attach();
    store.ui.attach();
    store.browserLocation.attach(fetchGraphql);

    new RouterListener().attach({
      router: Router,
      siteAnalytics: store.siteAnalytics,
      screenReaderSupport: store.screenReaderSupport,
    });
  }

  return store;
}
