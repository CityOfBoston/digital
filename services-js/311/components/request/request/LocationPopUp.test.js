// @flow

import React from 'react';
import renderer from 'react-test-renderer';
import { runInAction } from 'mobx';

import { AppStore } from '../../../data/store';
import RequestForm from '../../../data/store/RequestForm';
import LocationPopUp from './LocationPopUp';

const ACTIONS = {
  nextFunc: jest.fn(),
  nextIsSubmit: false,
};

describe('rendering', () => {
  let store;
  let requestForm;

  beforeEach(() => {
    store = new AppStore();
    store.apiKeys.mapbox = 'fake-api-key';

    requestForm = new RequestForm();
  });

  it('renders without an address', () => {
    runInAction(() => { requestForm.address = ''; });

    const component = renderer.create(
      <LocationPopUp store={store} requestForm={requestForm} {...ACTIONS} />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders with an address', () => {
    runInAction(() => { requestForm.address = 'City Hall Square\nBoston, MA'; });
    const component = renderer.create(
      <LocationPopUp store={store} requestForm={requestForm} {...ACTIONS} />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders a map on a small screen', () => {
    store.ui.visibleWidth = 320;
    const component = renderer.create(
      <LocationPopUp store={store} requestForm={requestForm} {...ACTIONS} />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
