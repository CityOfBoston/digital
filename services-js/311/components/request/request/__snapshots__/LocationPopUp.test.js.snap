// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders a map on a small screen 1`] = `
<div
  className={
    Object {
      "data-css-a1pmen": "",
    }
  }
>
  <div
    className="css-1quve7l"
  />
  <div
    className="p-a300 css-wi4u21"
  >
    <form
      action="#"
      className="sf sf--sm sf--y"
      onSubmit={[Function]}
    >
      <div
        className="sf-i"
      >
        <input
          aria-label="Address search field"
          className="sf-i-f"
          onInput={[Function]}
          placeholder="Search address or intersection…"
          type="text"
          value=""
        />
        <button
          className="sf-i-b"
          disabled={true}
          type="submit"
        >
          Search
        </button>
      </div>
    </form>
    <div
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <div
        className="m-v400"
      />
    </div>
    <div
      className="g css-1k8tu3j"
    >
      <div
        className="g--7 t--subinfo m-v200"
      >
        <a
          href="javascript:void(0)"
          onClick={[Function]}
        >
          Continue without location
        </a>
      </div>
      <button
        className="btn g--5"
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        Next
      </button>
    </div>
  </div>
</div>
`;

exports[`rendering renders with an address 1`] = `
<div
  className={
    Object {
      "data-css-a1pmen": "",
    }
  }
>
  <div
    className="p-a300 css-wi4u21"
  >
    <form
      action="#"
      className="sf sf--sm sf--y"
      onSubmit={[Function]}
    >
      <div
        className="sf-i"
      >
        <input
          aria-label="Address search field"
          className="sf-i-f"
          onInput={[Function]}
          placeholder="Search for a street address or intersection…"
          type="text"
          value=""
        />
        <button
          className="sf-i-b"
          disabled={true}
          type="submit"
        >
          Search
        </button>
      </div>
    </form>
    <div
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <div
        className="m-v400"
      >
        <div
          className="addr addr--s"
          style={
            Object {
              "whiteSpace": "pre-line",
            }
          }
        >
          City Hall Square
          Boston, MA
        </div>
      </div>
    </div>
    <div
      className="g css-1k8tu3j"
    >
      <div
        className="g--7 t--subinfo m-v200"
      >
        <a
          href="javascript:void(0)"
          onClick={[Function]}
        >
          Continue without location
        </a>
      </div>
      <button
        className="btn g--5"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Next
      </button>
    </div>
  </div>
</div>
`;

exports[`rendering renders without an address 1`] = `
<div
  className={
    Object {
      "data-css-a1pmen": "",
    }
  }
>
  <div
    className="p-a300 css-wi4u21"
  >
    <form
      action="#"
      className="sf sf--sm sf--y"
      onSubmit={[Function]}
    >
      <div
        className="sf-i"
      >
        <input
          aria-label="Address search field"
          className="sf-i-f"
          onInput={[Function]}
          placeholder="Search for a street address or intersection…"
          type="text"
          value=""
        />
        <button
          className="sf-i-b"
          disabled={true}
          type="submit"
        >
          Search
        </button>
      </div>
    </form>
    <div
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <div
        className="m-v400"
      />
    </div>
    <div
      className="g css-1k8tu3j"
    >
      <div
        className="g--7 t--subinfo m-v200"
      >
        <a
          href="javascript:void(0)"
          onClick={[Function]}
        >
          Continue without location
        </a>
      </div>
      <button
        className="btn g--5"
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        Next
      </button>
    </div>
  </div>
</div>
`;
